hail/python/hail/ir/ir.py

```diff
@@ -2865,6 +2866,65 @@ def renderable_bindings(self, i, default_value=None):
     def renderable_agg_bindings(self, i, default_value=None):
         return self.child.typ.entry_env(default_value) if i == 1 else {}
 
+class PartitionReader(object):
+    pass
+
+    def set_uid_field(self, uid_field):
+        pass
+
+    def render(self):
+        pass
+
+    def _eq(self, other):
+        pass
+
+
+class PartitionNativeIntervalReader(PartitionReader):
+    def __init__(self, path):
+        self.path = path
+        self.uid_field = '__dummy'
+
+    def set_uid_field(self, uid_field):
+        self.uid_field = uid_field
+
+    def render(self):
+        return escape_str(json.dumps({"name": "PartitionNativeIntervalReader",
+                "path": self.path,
+                "uidFieldName": self.uid_field}))
+
+    def _eq(self, other):
+        return isinstance(other, PartitionNativeIntervalReader) and self.path == other.path
+
+
+class ReadPartition(IR):
+    @typecheck_method(context=IR, row_typ=tstruct, reader=PartitionReader)
+    def __init__(self, context, row_typ, reader):
+        super().__init__(context)
+        self.context = context
+        self.row_typ = row_typ
+        self.row_typ_with_uid = row_typ
+        self.drop_uid = True
+        self.reader = reader
+
+    def _handle_randomness(self, create_uids):
+        if create_uids:
+            self.row_typ_with_uid = tstruct(**self.row_typ, __uid=ttuple(tint64, tint64))
+            self.drop_uid = False
+            self.reader.set_uid_field('__uid')
```

Comment:
All the other `_handle_randomness` methods return new nodes, they don't modify the node. I think that's necessary, since an IR can be reused in different contexts.